src/index/openid/cache.ts

```diff
@@ -24,74 +24,61 @@
  *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
  */
 
+/**
+ * @module openid/cache
+ */
+
 import { createHash } from 'crypto';
 
-import * as openidClient from 'openid-client';
-import { Environment } from '../..';
+import { Environment } from '..';
 
-/**
- * @private
- */
-const cache: { [s: string]: openidClient.Issuer; } = {};
+import { OpenIdClient } from './client';
 
-/**
- * @private
- */
-function createKey(httpTimeOut: number, openidIssuerUri: string) {
-	return createHash('sha1').update(openidIssuerUri + '|' + httpTimeOut).digest('hex');
+interface OpenIdClientCache {
+	[index: string]: OpenIdClient;
 }
 
 /**
- * @private
+ * Finds the OpenID client for the given environment.
+ *
+ * If the OpenID client is not found, a new client is created and stored in the cache.
+ *
+ * @param env The OpenID environment parameters.
  */
-async function buildIssuer(httpTimeOut: number, openidIssuerUri: string) {
+export async function findOrCreateOpenIdClient(env: Environment) {
 
-	openidClient.Issuer.defaultHttpOptions = {
-		timeout: httpTimeOut
-	};
+	const key = createKey(env);
 
-	return await openidClient.Issuer.discover(openidIssuerUri);
+	let client = cache[key];
+	if (!client) {
+		client = new OpenIdClient(env);
+		await client.init();
+		cache[key] = client;
+	}
+
+	return client;
 
 }
 
 /**
- * @private
+ * Clears the OpenID client cache.
+ *
+ * This results in the recreation of OpenID clients.
  */
-export function clearCache() {
+export function clear() {
 	Object.keys(cache).forEach((key) => {
 		delete cache[key];
 	});
 }
 
-/**
- * @private
- */
-export async function constructIssuer(env: Environment) {
-
-	const httpTimeOut = env.openidHTTPTimeout;
-	const openidIssuerUri = env.openidIssuerURI;
-	const key = createKey(httpTimeOut, openidIssuerUri);
-
-	let issuer = cache[key];
-	if (!issuer) {
-		issuer = await buildIssuer(httpTimeOut, openidIssuerUri);
-		cache[key] = issuer;
-	}
-
-	return Promise.resolve(issuer);
-
-}
+const cache: OpenIdClientCache = {};
```

Comment:
This could be replaced with the `Map` object. Think its a good candidate as you are clearing it and Map has a method for that already. It also enables the use of other types as keys (e.g. numbers) so you could implement a hashcode function to give you a true hashmap like Java instead of using crypto hashes, probably faster.

Reply:
Did a quick test with some hashcode funcs I found on stackoverflow:

```
-2061961108 ​​​​​

hash1: 0.125ms ​​​​​

-2061961108 ​​​​​

hash2: 0.066ms ​​​​​
 
be0cc9d4079dbb226343c784f7a349834ab4826b 

sha1: 0.058ms ​​​​​
```

yeah ignore the second part of the above comment..

Reply:
I'll look at rolling this into a subsequent change that I'm working on.